### DISCUSSION TITLE
The given address is not yours

### DISCUSSION THREAD

**Discussion by andreswatson** [QuorumDMS/ftp-srv#250] [Q&A] [answered]
Hi, I use this ftp-srv 4.4.0 module integrated into my App, once deployed in a server for staging I cannot figure out why I get an error "500 The given address is not yours" no matter Active Mode or Passive Mode. I deployed the ftp server in the PORT 2121, here the part of the communication:

02/04/2021 15:07:33  Sent> PORT 192,168,1,13,211,113
02/04/2021 15:07:33  Recv> 500 The given address is not yours
02/04/2021 15:07:33  Unable to load ftp://13223@scanntech.stage.guayerd.com:2121/: **The given address is not yours**

192.168.1.13 is my local ip address, yes, I'm behind a router, but I can connect to other FTP Servers in Internet in Active mode with no problem.

Any idea?

**Comment by matt-forster** [comment]
Hi Andre - you at receiving that error because the server is trying to bind it’s address to your local IP. The PORT command tells the server to use that address to listen for incoming connections, if your application is running on a computer that does not have `192.168.1.13` as it’s active address - it will throw that error.

The way you can mitigate this is to make sure you are passing in the `url` and the `pasv_url` options into the constructor, as listed [here](https://github.com/autovance/ftp-srv#url).

I can help you further if you post your config here and we can get to the bottom of it.

**Reply by andreswatson** [reply]
Thanks Matt for your prompt response. Locally works, the problem is when is deployed and I'm indeed behind NAT router.

Here you have the constructor:

//
const FtpServer = require('ftp-srv');

const server = new FtpServer({
  log: Logger.createLogger({
    name: 'ftpLogger',
    streams: [
      {
        level: 'trace',
        path: 'log/ftp.log',
        type: 'rotating-file',
        period: '1d', // daily rotation
        count: 3, // keep 3 back copies
      },
    ],
  }),
  url: `ftp://${process.env.FTP_URL}:${process.env.FTP_PORT}`,
  pasv_url: `${process.env.FTP_PASV_URL}`,
  pasv_min: process.env.FTP_PASV_MIN,
  pasv_max: process.env.FTP_PASV_MAX,
  file_format: 'ls',
  anonymous: false,
  greeting: ['Bienvenido al FTP!'],
});
//

The environment variables got:

  FTP_URL=localhost
  FTP_PORT=2121
  FTP_PASV_URL=localhost
  FTP_PASV_MIN=65500
  FTP_PASV_MAX=65515

LOGS using a FTP Client called BEYOND COMPARE, looks it works, this is the CLIENT log:

23/04/2021 14:02:08  Stat> Connected.
23/04/2021 14:02:08  Recv> 220-Bienvenido al FTP!
23/04/2021 14:02:08  Recv> 220 Ready
23/04/2021 14:02:08  Sent> USER 13223
23/04/2021 14:02:08  Recv> 331 Username okay, awaiting password
23/04/2021 14:02:08  Sent> PASS ********
23/04/2021 14:02:08  Recv> 230 User logged in, proceed
23/04/2021 14:02:08  Sent> FEAT
23/04/2021 14:02:09  Recv> 211-Extensions supported
23/04/2021 14:02:09  Recv>  UTF8
23/04/2021 14:02:09        211 End
23/04/2021 14:02:09  Sent> OPTS UTF8 ON
23/04/2021 14:02:09  Recv> 200 UTF8 encoding on
23/04/2021 14:02:09  Sent> TYPE A
23/04/2021 14:02:09  Recv> 200 Switch to "ascii" transfer mode.
23/04/2021 14:02:09  Sent> SYST
23/04/2021 14:02:09  Recv> 215 UNIX Type: L8
23/04/2021 14:02:09  Sent> TYPE A
23/04/2021 14:02:09  Recv> 200 Switch to "ascii" transfer mode.
23/04/2021 14:02:09  Sent> PWD
23/04/2021 14:02:09  Recv> 257 "/"
23/04/2021 14:02:09  Sent> REST 1
23/04/2021 14:02:10  Recv> 350 Restarting next transfer at 1
23/04/2021 14:02:10  Sent> REST 0
23/04/2021 14:02:10  Recv> 350 Restarting next transfer at 0
23/04/2021 14:02:10  Sent> PASV
23/04/2021 14:02:10  Recv> 227 PASV OK (localhost,21,124)
23/04/2021 14:02:10  Sent> LIST -a
23/04/2021 14:02:10  Recv> 150 File status okay; about to open data connection
23/04/2021 14:02:10  Recv> 226 Closing data connection
23/04/2021 14:02:38  Sent> TYPE I
23/04/2021 14:02:38  Recv> 200 Switch to "binary" transfer mode.
23/04/2021 14:02:38  Sent> PASV
23/04/2021 14:02:38  Recv> 227 PASV OK (localhost,21,125)
23/04/2021 14:02:38  Sent> STOR kk.pdf
23/04/2021 14:02:39  Recv> 150 File status okay; about to open data connection
23/04/2021 14:02:40  Recv> 226 /kk.pdf
23/04/2021 14:02:40  Sent> TYPE A
23/04/2021 14:02:40  Recv> 200 Switch to "ascii" transfer mode.
23/04/2021 14:02:40  Sent> PASV
23/04/2021 14:02:40  Recv> 227 PASV OK (localhost,21,126)
23/04/2021 14:02:40  Sent> LIST -a
23/04/2021 14:02:41  Recv> 150 File status okay; about to open data connection
23/04/2021 14:02:41  Recv> 226 Closing data connection
23/04/2021 14:02:41  Successfully copied 1 items.  Completed in 2,88 seconds.

But as soon as I use any other CLIENT like FileZilla, ftp or lftp (my customer uses lftp) I start receiving errors like "The given address is not yours" even having sent PASV commands.

Here using ftp at my terminal from home to the remote server:

% ftp localhost 2121
Connected to localhost.
220-Bienvenido al FTP!
220 Ready
Name (localhost:andreswatson): 13223
331 Username okay, awaiting password
Password: 
230 User logged in, proceed
ftp> ls
500 The given address is not yours
502 Command not allowed: LPRT
425 No connection established
ftp> 

At the Server logs you can see:

{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","level":10,"port":2121,"encoding":"utf8","message":"220-Bienvenido al FTP!","msg":"Reply","time":"2021-04-23T17:09:10.874Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","level":10,"port":2121,"encoding":"utf8","message":"220 Ready","msg":"Reply","time":"2021-04-23T17:09:10.874Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","level":10,"msg":"[ 'USER 13223' ]","time":"2021-04-23T17:09:19.750Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","directive":"USER","level":10,"command":{"directive":"USER","arg":"13223","flags":[],"raw":"USER 13223"},"msg":"Handle command","time":"2021-04-23T17:09:19.751Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","level":10,"port":2121,"encoding":"utf8","message":"331 Username okay, awaiting password","msg":"Reply","time":"2021-04-23T17:09:19.751Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","level":10,"msg":"[ 'PASS 27zsoJDmPvy' ]","time":"2021-04-23T17:09:26.349Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","directive":"PASS","level":10,"command":{"directive":"PASS","arg":"********","flags":[],"raw":"PASS 27zoJDmPvy"},"msg":"Handle command","time":"2021-04-23T17:09:26.350Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","level":10,"port":2121,"encoding":"utf8","message":"230 User logged in, proceed","msg":"Reply","time":"2021-04-23T17:09:26.565Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","level":10,"msg":"[ 'PORT 192,168,1,13,241,239' ]","time":"2021-04-23T17:09:29.451Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","directive":"PORT","level":10,"command":{"directive":"PORT","arg":"192,168,1,13,241,239","flags":[],"raw":"PORT 192,168,1,13,241,239"},"msg":"Handle command","time":"2021-04-23T17:09:29.452Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","directive":"PORT","level":50,"err":{"message":"The given address is not yours","name":"SocketError","stack":"SocketError: The given address is not yours\n    at /var/www/localhost/api/node_modules/ftp-srv/src/connector/active.js:33:15\n    at tryCatcher (/var/www/localhost/api/node_modules/bluebird/js/release/util.js:16:23)\n    at Promise._settlePromiseFromHandler (/var/www/localhost/api/node_modules/bluebird/js/release/promise.js:547:31)\n    at Promise._settlePromise (/var/www/localhost/api/node_modules/bluebird/js/release/promise.js:604:18)\n    at Promise._settlePromiseCtx (/var/www/localhost/api/node_modules/bluebird/js/release/promise.js:641:10)\n    at _drainQueueStep (/var/www/localhost/api/node_modules/bluebird/js/release/async.js:97:12)\n    at _drainQueue (/var/www/localhost/api/node_modules/bluebird/js/release/async.js:86:9)\n    at Async._drainQueues (/var/www/localhost/api/node_modules/bluebird/js/release/async.js:102:5)\n    at Immediate.Async.drainQueues [as _onImmediate] (/var/www/localhost/api/node_modules/bluebird/js/release/async.js:15:14)\n    at processImmediate (internal/timers.js:461:21)","code":500},"msg":"The given address is not yours","time":"2021-04-23T17:09:29.453Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","level":10,"port":2121,"encoding":"utf8","message":"500 The given address is not yours","msg":"Reply","time":"2021-04-23T17:09:29.453Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","level":10,"msg":"[ 'LPRT 4,4,192,168,1,13,2,241,239' ]","time":"2021-04-23T17:09:29.619Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","directive":"LPRT","level":10,"command":{"directive":"LPRT","arg":"4,4,192,168,1,13,2,241,239","flags":[],"raw":"LPRT 4,4,192,168,1,13,2,241,239"},"msg":"Handle command","time":"2021-04-23T17:09:29.620Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","level":10,"port":2121,"encoding":"utf8","message":"502 Command not allowed: LPRT","msg":"Reply","time":"2021-04-23T17:09:29.620Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","level":10,"msg":"[ 'LIST' ]","time":"2021-04-23T17:09:29.788Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","directive":"LIST","level":10,"command":{"directive":"LIST","arg":null,"flags":[],"raw":"LIST"},"msg":"Handle command","time":"2021-04-23T17:09:29.789Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","directive":"LIST","level":50,"err":{"message":"operation timed out","name":"TimeoutError","stack":"TimeoutError: operation timed out\n    at afterTimeout (/var/www/localhost/api/node_modules/bluebird/js/release/timers.js:46:19)\n    at Timeout.timeoutTimeout [as _onTimeout] (/var/www/localhost/api/node_modules/bluebird/js/release/timers.js:76:13)\n    at listOnTimeout (internal/timers.js:554:17)\n    at processTimers (internal/timers.js:497:7)"},"msg":"operation timed out","time":"2021-04-23T17:09:34.792Z","v":0}
{"name":"ftpLogger","hostname":"nodejs-s-1vcpu-1gb-nyc3-01","pid":292766,"id":"9e58f4ee-e15b-4d91-9f94-255de9c740e3","ip":"201.216.223.162","level":10,"port":2121,"encoding":"utf8","message":"425 No connection established","msg":"Reply","time":"2021-04-23T17:09:34.792Z","v":0}

//////

The weird part is WHY and WHAT is doing the Beyond Compare that could finally send the file using PASSIVE MODE, and in the same machine, any other client gets this error.

Any idea??

Thanks in advance!

**Comment by guayerd** [comment]
Sorry @matt-forster, just tried to supply the whole information. Let me know if I misunderstood your request.

**Reply by matt-forster** [reply]
It looks good (if you are the author), just haven't had time to dissect it.

One thing I did notice, is that your `PASV_URL` is set to `localhost`. When you are behind NAT - the address that is used must be the IP address that the client is interacting with, which would be the WAN-facing IP, not the localhost address.

**Comment by matt-forster** [comment, marked as answer]
It looks good, just haven't had time to dissect it.

One thing I did notice, is that your `FTP_PASV_URL` is set to `localhost`. When you are behind NAT - the address that is used must be the IP address that the client is interacting with, which would be the WAN-facing IP, not the localhost address. That will 100% cause the error you are receiving - as to why Beyond doesn't get the error, I have no idea. It could be because it actually supports the `localhost` hostname and does not translate it to an IP.

**Reply by guayerd** [reply]
Thanks @matt-forster for your response. I changed the URL to publish here, at `FTP_PASV_URL` you have the server's [Fully qualified domain name](https://en.wikipedia.org/wiki/Fully_qualified_domain_name).

If I try to connect using a terminal ftp command, I get an error "The given address is not yours" even if a set the command `quote PASV` but I found if I set the command at ftp `quote EPSV` that refers to Extended Passive Mode, it works.

The problem is my customers don't have that option and they are experimenting problems with all kind of FTP Clients like Filezilla.

Any idea what else could be?

Thanks in advance!

**Reply by matt-forster** [reply]
Hmm, we certainly use our production version with Filezilla with passive mode - I'll check it out further.

**Reply by guayerd** [reply]
I've just tried with Filezilla:

```
Status:      	Logged in
Status:      	Retrieving directory listing...
Command:	PWD
Response: 	257 "/"
Command:	TYPE I
Response: 	200 Switch to "binary" transfer mode.
Command:	PASV
Response: 	227 PASV OK (scanntech,stage,guayerd,com,21,129)
Command:	PORT 192,168,1,13,226,31
Response: 	500 The given address is not yours
Error:        	Failed to retrieve directory listing
```

Look it pass the PASV mode command, and it's accepted but then the Filezilla reports the port with a local network / NAT IP address and gets an error `500 The given address is not yours`

In your production environment you mentioned, someone is behind a NAT / Router? I think more than 80% of the people is behind a router in a NAT local network.

**Reply by matt-forster** [reply]
Ok - so if I understand correctly, you've set the `PASV_URL` to a DNS name, and the response tp PASV in your reply returns that domain - when Filezilla sends the corresponding PORT command, it is using a local IP address. 

> Response: 	227 PASV OK (scanntech,stage,guayerd,com,21,129)
> Command:	PORT 192,168,1,13,226,31
> Response: 	500 The given address is not yours

It doesn't seem like the FTP clients like having a domain returned in response to the PASV command. __To me, it doesn't seem like the servers fault that the client is sending a local IP address for the server to connect to during the PORT command;__

From https://cr.yp.to/ftp/retr.html:

> The PORT verb
> A PORT request asks the server to use a different mechanism of creating a data connection: the server makes a TCP connection to the client.
> The PORT request has a parameter in the form
>     h1,h2,h3,h4,p1,p2
> meaning that the client is listening for connections on TCP port p1*256+p2 at IP address h1.h2.h3.h4. (The RFC 959 formal syntax does not allow any of these numbers to be 0. The formal syntax is wrong.)
> The server normally accepts PORT with code 200. If the server was listening for a connection, it stops, and drops any connections already made.

Our server has it's external IP address set instead of a domain name for the PASV URL, and that works - the spec doesn't seem to list DNS names as supported here.

**Reply by andreswatson** [reply]
Thanks @matt-forster. That was the mess. I proposed a change in the Docs to specify that must be an IP Address not a host https://github.com/autovance/ftp-srv/pull/253

Thanks a lot for your promptly response. Now is working.
